ludwig/schema/encoders/image_encoders.py

```diff
@@ -284,368 +284,828 @@ class Stacked2DCNNEncoderConfig(BaseEncoderConfig):
-        parameter_metadata=ENCODER_METADATA["ViTEncoder"]["patch_size"],
+    model_variant: Optional[str] = schema_utils.StringOptions(
+        [
+            "b_16",
+            "b_32",
+            "l_16",
+            "l_32",
+            "h_14",
+        ],
+        default="b_16",
+        allow_none=False,
+        description="Pretrained model variant to use.",
+        parameter_metadata=ENCODER_METADATA["TVViTEncoder"]["model_variant"],
     )
 
-    saved_weights_in_checkpoint: bool = schema_utils.Boolean(
-        default=False,
-        description="Are the pretrained encoder weights saved in this model's checkpoint? Automatically set to"
-        "True for trained models to prevent loading pretrained encoder weights from model hub.",
-        parameter_metadata=ENCODER_METADATA["ViTEncoder"]["saved_weights_in_checkpoint"],
+
+@register_encoder_config("vgg_torch", IMAGE)
+@dataclass
+class TVVGGEncoderConfig(TVBaseEncoderConfig):
+    type: str = schema_utils.StringOptions(
+        ["vgg_torch"],
+        default="vgg_torch",
+        allow_none=False,
+        description="Type of encoder.",
     )
 
-    trainable: bool = schema_utils.Boolean(
-        default=True,
-        description="Is the encoder trainable.",
-        parameter_metadata=ENCODER_METADATA["ViTEncoder"]["trainable"],
+    model_variant: Union[int, str] = schema_utils.OneOfOptionsField(
+        default=11,
+        description="Pretrained model variant to use.",
+        field_options=[
+            schema_utils.IntegerOptions(
+                [
+                    11,
+                    13,
+                    16,
+                    19,
+                ],
+                default=11,
+                allow_none=False,
+            ),
+            schema_utils.StringOptions(
+                [
+                    "11_bn",
+                    "13_bn",
+                    "16_bn",
+                    "19_bn",
+                ],
+                default="11_bn",
+                allow_none=False,
+            ),
+        ],
+        parameter_metadata=ENCODER_METADATA["TVVGGEncoder"]["model_variant"],
     )
 
-    use_pretrained: bool = schema_utils.Boolean(
-        default=True,
-        description="Use pre-trained model weights from Hugging Face.",
-        parameter_metadata=ENCODER_METADATA["ViTEncoder"]["use_pretrained"],
+
+@register_encoder_config("wide_resnet_torch", IMAGE)
+@dataclass
+class TVWideResNetEncoderConfig(TVBaseEncoderConfig):
+    type: str = schema_utils.StringOptions(
+        ["wide_resnet_torch"],
+        default="wide_resnet_torch",
+        allow_none=False,
+        description="Type of encoder.",
     )
 
-    pretrained_model: str = schema_utils.String(
-        default="google/vit-base-patch16-224",
-        description="The name of the pre-trained model to use.",
-        parameter_metadata=ENCODER_METADATA["ViTEncoder"]["pretrained_model"],
+    model_variant: Optional[str] = schema_utils.StringOptions(
```

Comment:
Want to double-check if this field is really `Optional`? If so, make sure `allow_none` is `True`.

Reply:
@ksbrar   I want to clarify the feedback.  

For the torchvision encoders, the following should be legal syntax

```
encoder:
     type: efficientnet_torch
    model_variant: "b1"
```

For this second case, the `model_variant` parameter should take the default value, i.e., "b0".
```
encoder:
    type: efficientnet_torch
```


The following should be flagged as an error

```
encoder:
    type: efficientnet_torch
   model_variant: None
```

I'm thinking that `Optional[str]` for `model_variant` indicates that if I don't specify the `model_variant` parameter, the default value will be used.   And that `allow_none=False` says that `None` cannot be specified as a value for `model_variant`.

Am I interpreting the specifications correctly?

Reply:
@jimthompson5802 Technically `Optional[str]` is equivalent to setting `allow_none=True` - that is:
```
encoder:
    type: efficientnet_torch
   model_variant: None
```
would be accepted without error.

Some further context:

The reason that it doesn't matter functionally for us is because the custom marshmallow fields we've defined like `StringOptions` are in control of their own JSON schemas. In a less complicated world we would not have any custom fields and instead use native marshmallow fields and in that case the type annotation itself would dictate that field's JSON schema. So `Optional` usage would really matter.

In the Ludwig mental model, essentially every field has a default value so if the user fails to specify a value for a certain parameter then *something* is picked (even `None` in certain cases). Conversely, to make a parameter *required* (which is very, very rare)  the schema would only declare the parameter but then not set it to anything. So for instance if you wanted to make `model_variant` required the schema would look like:

```
@dataclass
class TVEncoder(...)
    type: str = ProtectedString(...) <--- has a default argument

    model_variant: str <-- has no default, is not set by the schema, is required to be set by the user
```

Does that make sense?

Reply:
Yep...the explaination helped.   Removed the Optional designation.